### PR TITLE
2512 feat/add property display setting

### PR DESCRIPTION
**Pull request by Ryoken**
#### What's this PR do?
Adds an organization setting to change which field is used to display a property's name.

#### How should this be manually tested?
See the org settings page, change the setting and then view a property detail page.

#### What are the relevant tickets?
https://github.com/SEED-platform/seed/issues/2512

#### Screenshots (if appropriate)
(see ticket)

### CODE REVIEW

File: seed/static/seed/js/controllers/inventory_detail_analyses_controller.js [hunk 1]
```diff
@@ -103,6 +103,14 @@ angular.module('BE.seed.controller.inventory_detail_analyses', [])
         })
       }
 
+      $scope.inventory_display_name = function() {
+        let field = $scope.org.property_display_field;
+        if (!$scope.item_state[field]) {
+          field = 'address_line_1';
+        }
+        return $scope.item_state[field] ? $scope.item_state[field] : '(invalid display field)';
```
Comment (macintoshpie) [inline seed/static/seed/js/controllers/inventory_detail_analyses_controller.js:111]: I like what you've got here, very clean. I'm wondering if this message might confuse some folks who don't manage the org settings. What do you think about something like `(no ${field})`, or even better `(no ${display name for the field})`, kind of like SEED displayed before?
Also I'm wondering if there is a way we could help org managers fix this if they see this message, ie some help message or link to settings. Not sure if either option is good UI, but in general its useful to say how to fix things in addition to reporting a problem

**Review by macintoshpie — COMMENTED** [review]
Very nice! I left a few comments/questions

File: seed/static/seed/js/services/inventory_service.js [hunk 2]
```diff
@@ -498,6 +498,21 @@ angular.module('BE.seed.service.inventory', []).factory('inventory_service', [
...
+    inventory_service.get_property_column_names_for_org = function (org_id) {
+      return $http.get('/api/v3/columns/', {
+        params: {
+          inventory_type: 'property',
+          organization_id: org_id,
+          only_used: false,
+          display_units: true
+        }
+      }).then(function (response) {
+        let names = Array.from(new Set(response.data.columns.map(a => a.column_name)));
```
[the remark this answers is not in the document]
Comment (Ryoken) [reply seed/static/seed/js/services/inventory_service.js:510]: Yeah, here's an output of "response.data.columns.map(a => a.column_name)" showing duplicates.
Screen Shot 2021-01-27 at 4 44 17 PM

It looks like they're due to belonging to multiple tables.
Screen Shot 2021-01-27 at 4 46 46 PM

File: seed/static/seed/partials/inventory_detail.html [hunk 3]
```diff
@@ -20,8 +20,8 @@ <h1>{$:: (inventory_type === 'taxlots' ? 'Tax Lot Detail' : 'Property Detail') |
             <div class="section_action_container left" style="width: 50%;">
                 <span>
                     <h2>
-                        <span ng-if="::inventory_type==='properties'"><i class="fa fa-building-o"></i><span translate>Property</span> : {$ item_state.address_line_1 ? item_state.address_line_1 : ('(no address 1)' | translate) $}</span>
-                        <span ng-if="::inventory_type==='taxlots'"><i class="fa fa-map-o"></i><span translate>Tax Lot</span> : {$ item_state.address_line_1 ? item_state.address_line_1 : ('(no address 1)' | translate) $}</span>
+                        <span ng-if="::inventory_type==='properties'"><i class="fa fa-building-o"></i><span translate>Property</span> : {$ inventory_display_name() | translate $}</span>
+                        <span ng-if="::inventory_type==='taxlots'"><i class="fa fa-map-o"></i><span translate>Tax Lot</span> : {$ inventory_display_name() | translate) $}</span>
```
[the remark this answers is not in the document]
Comment (Ryoken) [reply seed/static/seed/partials/inventory_detail.html:24]: The previous implementation uses "address_line_1" regardless of if it was taxlot or property, so I didn't think a way to specify was needed.  I can definitely put in a second setting for tax lots and pull them seperately though.

File: seed/lib/superperms/orgs/models.py [hunk 4]
```diff
@@ -193,6 +193,8 @@ class Meta:
 
     geocoding_enabled = models.BooleanField(default=True)
 
+    property_display_field = models.CharField(max_length=32, blank=False, default="address_line_1")
```
[the remark this answers is not in the document]
Comment (Ryoken) [reply seed/lib/superperms/orgs/models.py:196]: Yeah, it felt simpler at the cost of very rarely not having a correct lookup string, at which point the default kicks in just fine.

File: seed/static/seed/js/controllers/inventory_detail_analyses_controller.js (hunk 1 above)
Comment (Ryoken) [reply seed/static/seed/js/controllers/inventory_detail_analyses_controller.js:111]: I was actually thinking about putting in something along those lines so I'm glad you brought it up too.  I'll update with something more clear and possibly a tooltip to direct users to fix it.

File: seed/static/seed/partials/inventory_detail.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (adrian-lara) [reply seed/static/seed/partials/inventory_detail.html:24]: Ah good catch here @macintoshpie . Yeah, we should probably have a separate setting for each.

File: seed/static/seed/js/services/inventory_service.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (macintoshpie) [reply seed/static/seed/js/services/inventory_service.js:510]: Yikes, now that I think about this, we shouldn't be getting TaxLotState results if we provide 'property' as the inventory type. Would you mind digging into this to see if that's what's actually happening?

File: seed/static/seed/js/controllers/inventory_detail_analyses_controller.js (hunk 1 above)
Comment (Ryoken) [reply seed/static/seed/js/controllers/inventory_detail_analyses_controller.js]: screencapture

File: seed/static/seed/js/services/inventory_service.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (macintoshpie) [reply seed/static/seed/js/services/inventory_service.js]: Ah ok then I guess the API is doing what it says it should, I suppose we will filter on the frontend. Yeah, glancing at that it does seem strange, but maybe I don't understand really what the `related` thing does?

File: seed/static/seed/js/controllers/inventory_detail_analyses_controller.js (hunk 1 above)
Comment (macintoshpie) [reply seed/static/seed/js/controllers/inventory_detail_analyses_controller.js]: ohh looks very nice!

File: seed/static/seed/js/services/inventory_service.js [hunk 5]
```diff
@@ -498,6 +498,37 @@ angular.module('BE.seed.service.inventory', []).factory('inventory_service', [
...
+        }
+      }).then(function (response) {
+        let property_columns = response.data.columns.filter(column => column.table_name == 'PropertyState');
+        return property_columns.map(a => {
+          return { 'column_name': a.column_name, 'display_name': property_columns.find(x => x.column_name == a.column_name).display_name };
+        });
+      });
+    };
+
+    inventory_service.get_taxlot_column_names_for_org = function (org_id) {
```
Comment (macintoshpie) [inline seed/static/seed/js/services/inventory_service.js:517]: I didn't realize this before, but these are very similar to the `get_property_columns`/`get_property_columns_for_org` (same for taxlot) functions, not sure what the best option is here. I guess we could modify the original func with `ui_grid_formatting = true` then conditionally format the display name? Or are there other differences I'm missing as well?

File: seed/static/seed/js/services/inventory_service.js [hunk 6]
```diff
@@ -498,6 +498,37 @@ angular.module('BE.seed.service.inventory', []).factory('inventory_service', [
...
+        params: {
+          inventory_type: 'property',
+          organization_id: org_id,
+          only_used: false,
+          display_units: true
+        }
+      }).then(function (response) {
+        let property_columns = response.data.columns.filter(column => column.table_name == 'PropertyState');
+        return property_columns.map(a => {
+          return { 'column_name': a.column_name, 'display_name': property_columns.find(x => x.column_name == a.column_name).display_name };
```
Comment (macintoshpie) [inline seed/static/seed/js/services/inventory_service.js:512]:
```suggestion
          return { 'column_name': a.column_name, 'display_name': a.display_name };
```
Would this work? Same goes for taxlot func

File: seed/static/seed/js/controllers/inventory_detail_analyses_controller.js [hunk 7]
```diff
@@ -103,6 +103,19 @@ angular.module('BE.seed.controller.inventory_detail_analyses', [])
...
       }
 
+      $scope.inventory_display_name = function(property_type) {
+        let error = '';
+        let field = property_type == "property" ? $scope.org.property_display_field : $scope.org.taxlot_display_field;
+        if (!(field in $scope.item_state)) {
+          error = field + ' does not exist';
+          field = 'address_line_1';
+        }
+        if (!$scope.item_state[field]) {
```
Comment (macintoshpie) [inline seed/static/seed/js/controllers/inventory_detail_analyses_controller.js:113]: Looks like I'm getting both messages:
Screen Shot 2021-01-29 at 1 51 58 PM

File: seed/static/seed/partials/inventory_detail.html [hunk 8]
```diff
@@ -20,8 +20,13 @@ <h1>{$:: (inventory_type === 'taxlots' ? 'Tax Lot Detail' : 'Property Detail') |
...
                     <h2>
-                        <span ng-if="::inventory_type==='properties'"><i class="fa fa-building-o"></i><span translate>Property</span> : {$ item_state.address_line_1 ? item_state.address_line_1 : ('(no address 1)' | translate) $}</span>
-                        <span ng-if="::inventory_type==='taxlots'"><i class="fa fa-map-o"></i><span translate>Tax Lot</span> : {$ item_state.address_line_1 ? item_state.address_line_1 : ('(no address 1)' | translate) $}</span>
+                        <span ng-if="::inventory_type==='properties'">
+                            <i class="fa fa-building-o"></i><span translate>Property</span> : 
+                        </span>
+                        <span ng-if="::inventory_type==='taxlots'">
+                            <i class="fa fa-map-o"></i><span translate>Tax Lot</span> : 
+                        </span>
+                        <span ng-bind-html="inventory_name">{$ inventory_display_name(inventory_type==='properties' ? 'property' : 'taxlot') | translate $}</span>
```
Comment (macintoshpie) [inline seed/static/seed/partials/inventory_detail.html:29]: Looks like sanitization is working properly 😉 
Screen Shot 2021-01-29 at 2 21 07 PM
Screen Shot 2021-01-29 at 2 23 13 PM

File: seed/views/v3/organizations.py [hunk 9]
```diff
@@ -516,6 +518,16 @@ def warn_bad_units(kind, unit_string):
         if geocoding_enabled != org.geocoding_enabled:
             org.geocoding_enabled = geocoding_enabled
 
+        # Update property_display_field option
+        property_display_field = posted_org.get('property_display_field', True)
```
Comment (macintoshpie) [inline seed/views/v3/organizations.py:522]:
```suggestion
        property_display_field = posted_org.get('property_display_field', '')
```
or maybe None, then don't update if None?

File: seed/views/v3/organizations.py [hunk 10]
```diff
@@ -1310,3 +1322,25 @@ def geocoding_enabled(self, request, pk=None):
...
+        return org.property_display_field
+
+    @has_perm_class('requires_member')
+    @ajax_request_class
+    @action(detail=True, methods=['GET'])
+    def taxlot_display_field(self, request, pk=None):
+        """
+        Returns the organization's taxlot_display_field setting
+        """
+        org = Organization.objects.get(id=pk)
```
Comment (macintoshpie) [inline seed/views/v3/organizations.py:1344]: Are these being used anywhere? I didn't see it anywhere, but could have missed it. Wondering if these endpoints are necessary

**Review by macintoshpie — COMMENTED** [review]
Tested locally and everything is working as expected except for the case where the inventory doesn't have the requested column and it's address_line_1 is empty (see comment). Few comments/questions on code but things are looking good

File: seed/static/seed/js/services/inventory_service.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ryoken) [reply seed/static/seed/js/services/inventory_service.js]: Confirmed that the tables aren't filtered out, rather tagged with the "related" column.  Working as expected!

File: seed/static/seed/js/services/inventory_service.js (hunk 6 above)
Comment (Ryoken) [reply seed/static/seed/js/services/inventory_service.js:512]: Actually, now that we are intelligently removing the duplicates, yeah, that would work much nicer.

File: seed/views/v3/organizations.py (hunk 10 above)
Comment (Ryoken) [reply seed/views/v3/organizations.py:1344]: No, they are not being used, I just assumed that new members needed to be added.  I honestly don't know if they will be super usefull either so I can remove them.

File: seed/static/seed/js/controllers/inventory_detail_analyses_controller.js (hunk 7 above)
Comment (Ryoken) [reply seed/static/seed/js/controllers/inventory_detail_analyses_controller.js:113]: Modified to accomodate.

File: seed/views/v3/organizations.py (hunk 9 above)
Comment (Ryoken) [reply seed/views/v3/organizations.py:522]: Updated to have the correct default "address_line_1", thanks for the catch.

File: seed/static/seed/js/services/inventory_service.js (hunk 5 above)
Comment (Ryoken) [reply seed/static/seed/js/services/inventory_service.js:517]: This endpoint is returning an array of tuples, basically, with only the column name and the display name.  This tuple array is used to populate the select from the org settings page.  Do you think the original functions could and should be used for this?

Comment (macintoshpie) [reply seed/static/seed/js/services/inventory_service.js]: I'm good with just keeping it how you have it now, easy enough and your functions are simple.

**Pull request merged**